docs/Contributing/Blueprints/Troubleshooting-Hosts.md

```diff
@@ -0,0 +1,195 @@
+- To not generate more load we will have a new column in `hosts` called something like `trouble_queries`.
+Given that we load the host by selecting `hosts` on every request, this should cause no extra performance penalty on every request).
+Alternatively we could use Redis to store such state, but it would add a Redis read request on every `distributed/read` request.
+- We need to use and store the hashes of the queries because queries can change and new queries be
+  added in between read and write requests (e.g. policy queries are editable by users and new ones can be assigned to hosts).
+
+#### `distributed/read`
+	
+```go
+func determineQueriesToSend(host *fleet.Host, queriesToSend map[string]string) map[string]string {
+	hostNotResponding := svc.detectHostNotResponding(host)
+	switch {
+
+	case !hostNotResponding && host.TroubleQueries == "":
+
+		// Nothing to do, host is responding and there are no "trouble queries".
+		return queriesToSend
+
+	case hostNotResponding && host.TroubleQueries == "":
+
+		// This is the first time Fleet detects the host has issues.
+		
+		// We assume hostDetailQueries never have performance issues.
+		nonDetailQueries := excludeQueries(queriesToSend, hostDetailQueries)
+		
+		host.TroubleQueries = nonDetailQueries // mark all to be sent as "trouble queries"
+		svc.ds.UpdateHost(ctx, host)
+		queriesToSend := excludeQueries(queriesToSend, host.TroubleQueries)
+		addQueries(queriesToSend, host.TroubleQueries[0:len(queriesToSend)/2]) // send first half of queries
+		
+		// generateTroubleQuery generates a string of the form: SELECT "q0:hash(q0),q1:hash(q1),..."
+		// which serves as an indicator of the sent queries on distributed/write.
+		queriesToSend["trouble_queries"] = generateTroubleQuery(nonDetailQueries)
+
+		return queriesToSend
+
+	case !hostNotResponding && host.TroubleQueries != "":
+
+		// This means the host is in "troubleshooting mode" but it has responded to some queries.
+	
+		queriesToSend := excludeQueries(queriesToSend, host.TroubleQueries)
+		troubleQueries := host.TroubleQueries[0:len(host.TroubleQueries/2)] // send first half of queries
+		addQueries(queriesToSend, troubleQueries)
+		
+		// generateTroubleQuery generates a string of the form: SELECT "q0:hash(q0),q1:hash(q1),..."
+		// which serves as an indicator of the sent queries on distributed/write.
+		queriesToSend["trouble_queries"] = generateTroubleQuery(queriesToSend)
+
+	case hostNotResponding && host.TroubleQueries != "":
+		
+		// This means the host is in "troubleshooting mode" and hasn't responded to first half of trouble queries.
+		
+		queriesToSend := excludeQueries(queriesToSend, host.TroubleQueries)
+		troubleQueries := host.TroubleQueries[len(host.TroubleQueries)/2: len(host.TroubleQueries)] // send second half of queries
+		addQueries(queriesToSend, troubleQueries)
+
+		// generateTroubleQuery generates a string of the form: SELECT "q0:hash(q0),q1:hash(q1),..."
```

Comment:
How about we add a batch size and batch index to the `q0:...` string? When we detect an issue, we batch by `len(queriesToSend)/2` and batch index of 0. So we send the `queriesToSend[batchSize*batchIndex:batchSize*batchIndex+1]`?

You then increase the `batchIndex` in 1 when `!hostNotResponding` and otherwise you reset the index to 0 and decrease the `batchSize` in 1.

Reply:
The "trouble query" is only **produced** in `distributed/read`, and can only be **accessed** in `distributed/write`.
In other words, `distributed/read` doesn't have access to such query result, it can only produce one.

Let me know if it makes sense.

Reply:
Ah, your idea makes sense now, added a rough approximation to the algorithm :)

Reply:
Basically I now defined the state as:
```go
type TroubleQueriesState struct {
	BatchIndex int
	BatchSize  int
	// Queries holds the suspicious queries (currently in search for culprits).
	Queries    []QueryWithHash
	// Denylisted holds the queries that have been confirmed to cause issues in the host.
	Denylisted []QueryWithHash
}
```